### DISCUSSION TITLE
What is a non-filtering sampler?

### DISCUSSION THREAD

**Discussion by TheMostDiligent** [gpuweb/gpuweb#4726] [Q&A] [answered]
The spec does not say anything about what this kind of sampler is supposed to do

**Comment by greggman** [comment, marked as answer]
"non-filtering" is a sampler binding type on a group binding layout entry

When you bind a sampler to a bind group via `createBindGroup`, there are validation rules in [section 8.2.1 Bind Group Creation](https://gpuweb.github.io/gpuweb/#bind-group-creation)

Those rules effectively say, if the bind group layout entry specifies "non-filtering", then the sampler being bound must use only 'nearest' filtering

**Reply by TheMostDiligent** [reply]
So if one of the min/mag filters is 'nearest', but the other is not, the sampler is still 'filtering'?

**Reply by greggman** [reply]
correct, minFilter, magFilter, and mipmapFilter must all be "nearest" to be "non-filtering". 

Note: I think I agree this is not explicit. It's implicit and maybe should be made explicit. It currently says

> [[isFiltering]], of type [boolean](https://webidl.spec.whatwg.org/#idl-boolean)
Whether the [GPUSampler](https://gpuweb.github.io/gpuweb/#gpusampler) weights multiple samples of a texture. from [section 7.1](https://gpuweb.github.io/gpuweb/#gpusampler). It should probably say exactly what settings = "weights multiple samples"

**Reply by TheMostDiligent** [reply]
Is there a reason for having such an unusual requirement in the API?
I see how filtering vs comparison samplers are different, but I can't see the need to specifically identify 'nearest'/'nearest'/'nearest' filter on the host.

**Reply by greggman** [reply]
Some textures are not filterable. They require a non-filtering sampler. Off the top of my head, integer based textures. Also, float32 textures until your device supports and you requires the "float32-filterable" feature

**Reply by TheMostDiligent** [reply]
TBH, filterable vs unfilterable textures and filtering vs non-filtering samplers are extremely confusing concepts that are not found in any other graphics APIs.
Textures are just samples of data. Whether this data can be filtered on not is the property of the device, not the data itself.

**Reply by greggman** [reply]
I think the point is to pre-validate as much as possible. If this didn't exist and you could just bind any sampler then this stuff would have to be checked at draw time. With this design, these restrictions can be checked at bind time and then there's less to check at draw time.

Other APIs most certainly do care about this stuff. They either have undefined behavior if done wrong, or they emit errors.

**Reply by TheMostDiligent** [reply]
Validation can happen when textures and samplers are set in the bind group, not in the draw time.

**Reply by greggman** [reply]
There's nothing that associates which samplers will be used with which textures at bind time, nor which bindGroups will be bound. What there is is that a certain pipeline needs bindGroups with certain layouts, each bindGroup was created with a certain layout. So at draw/dispatch time, if the bound bind groups' layouts match the pipeline's required layouts there's nothing left to check (or less to check). This check is cheap. A signature for layouts can be generated at creation time so to check it's as simple as comparing signature pointers. If they match, then the bind group is compatible with the pipeline, nothing left to check.

**Reply by TheMostDiligent** [reply]
And for unfilterable textures, the idea is that on a device that does not support f32 filtering, one defines an unfilterable texture slot in the bind group where an f32 texture will be bound?

**Reply by greggman** [reply]
You would need a unfilterable-float texture slot and, if you need a sampler, a non-filtering sampler slot

* [this example uses a 2x2 rgba8unorm texture](https://jsgist.org/?src=b33a27993c4fac4c2a703d1175e4793c)

  It uses `layout: 'auto'` when creating the pipeline

* [this example uses a 2x2 rgba32float texture](https://jsgist.org/?src=3336f95c8021e0cb0a160329e67ae05d)

  To get it to work, instead of using `layout: 'auto'` with the pipeline, it manually creates a bindGroupLayout

  ```js
  const bindGroupLayout = device.createBindGroupLayout({
    label: 'unfilterable-bgl',
    entries: [
      {
        binding: 0,
        visibility: GPUShaderStage.VERTEX,
        buffer: {
          minBindingSize: 128,
        },
      },
      {
        binding: 1,
        visibility: GPUShaderStage.FRAGMENT,
        buffer: {
          minBindingSize: 16,
        },
      },
      {
        binding: 2,
        visibility: GPUShaderStage.FRAGMENT,
        sampler: {
          type: "non-filtering",              // <---------
        },
      },
      {
        binding: 3,
        visibility: GPUShaderStage.FRAGMENT,
        texture: {
          sampleType: "unfilterable-float",   // <---------
        },
      },
    ],
  });
  ```

  Manually creates a pipeline layout

  ```js
  const pipelineLayout = device.createPipelineLayout({
    bindGroupLayouts: [bindGroupLayout],
  });
  ```

  And uses that when creating the pipeline

  ```js
  const pipeline = device.createRenderPipeline({
    layout: pipelineLayout,
    ...
  });
  ```

Note: I used the [WGSL offset generator](https://webgpufundamentals.org/webgpu/lessons/resources/wgsl-offset-computer.html#x=5d000001009f05000000000000003d888b0237284c230f76f4c740afd5a8274a453757e7450bcd700ba09ff07f29669c86d42c85e89b4005ad013c0308cf714c4ca5811fd63a1bddb6e51ae325d7eb55454104c323d48d5990d97136fc6aa2d95d9dad3fc538c3d3b18abfcc05f68a1605ac86c565629b94bdf5e913a77c9fa28b2b260656f0a823f31c213282ecfa4bb980b2386fb6d27e51c0b32a860d0444187bbaaee8833d9186e5d89b49a0739373dddb14ddfe96090153b8ac2affbae09747e09d7820b05fd050d7aea55e2ab4e5b435204e3ae302b6c20673fe4e56a1636875dd8606d4e76603a40eb944e9d274f7dc763ca679d587bfc8db74a2aa2194dd5da75e6a988b04d4019bc06e0e188cf3c647a69ca86d0db08e74724a2e65e2882c2183d9d7ba5af995670db9ee0f58cb5ebc9e2055c683e0f2426bb16d1e6bbc1d3b8eed5dd761b0e1ead393c28315c00c963175fae038a8a84a919f59593857c1984bcee55fc9bcd005447568d0bd2e694223ad309ea180df8a0c2e1bf712b56780335f1e6efcc872b58d4ca5284713230aafe6df7ddacfb2de5344ee88ff54412d3ce78943f189c3e0f88114c80c5571990b1dfbf30918ca6f38cb4fe2d68fa1780d48cfb9074cd12c4c78025e2d37dc1fef012b879eefbc60f7caa34483ff7e1a099b0fdf58efe46bc01d13e13bcfd1e60c0aa627bd65012c8647e1b92140e40e71e6331cd2751e1f2df0ba22bdbefa80a94a0e0d146f907537e523d863b8c98804668226818177e68dde26df4846ffff55045500), pasting in the shaders and it generated a bind group layout for me with the same settings as `layout: 'auto'`. I then changed the texture entry to "unfiltered-float" and sampler entry "non-filtering"

**Reply by TheMostDiligent** [reply]
Thank you for the explanation!

**Reply by TheMostDiligent** [reply]
BTW, is it allowed to bind 'nearest'/'nearest'/'nearest' filter to a filtering sampler?

**Reply by greggman** [reply]
yes, that sample above with the 2x2 rgba8unorm texture is using nearest/nearest/nearest

**Reply by kainino0x** [reply]
> Is there a reason for having such an unusual requirement in the API?

Yes. In other APIs it is undefined behavior (undefined results, could be uninitialized memory) if you attempt to filter a non-filterable texture. In WebGPU we can't allow that for security reasons.

(EDIT: and as gregg said we strongly want to validate earlier instead of validating at the last possible moment, which would also be safe)

**Reply by TheMostDiligent** [reply]
Yes, I understand that WebGPU is designed around safety. It is just that I would personally prefer the API to be more convenient to use at the price of a minor run-time performance hit.

Thanks for the explanation!

**Reply by greggman** [reply]
> I would personally prefer the API to be more convenient to use at the price of a minor run-time performance hit.

It's not a minor performance hit. Since you can bind 16 samplers and 16 textures as well as uniform buffers, storage buffers, and storage textures, all of which have requirements, and you can do 10s of thousands of draw calls per frame, it's saving up to several million checks per frame to do it early.

You could maybe cache that bindGroupX worked on slotN of pipelineM on the first draw but that would still add overhead to look up whether that a particular bindGroup had been checked for usage on a particular group slot of a particular pipeline.
